service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+		Service: request.Service,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &operatorservice.CreateOrUpdateNexusIncomingServiceResponse{Service: entryToService(resp.Entry)}, nil
+}
+
+func (r *registry) DeleteNexusIncomingService(
+	ctx context.Context,
+	request *operatorservice.DeleteNexusIncomingServiceRequest,
+) (*operatorservice.DeleteNexusIncomingServiceResponse, error) {
+	_, err := r.matchingClient.DeleteNexusIncomingService(ctx, &matchingservice.DeleteNexusIncomingServiceRequest{
+		Name: request.Name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &operatorservice.DeleteNexusIncomingServiceResponse{}, nil
+}
+
+func (r *registry) GetNexusIncomingService(
+	ctx context.Context,
+	request *operatorservice.GetNexusIncomingServiceRequest,
+) (*operatorservice.GetNexusIncomingServiceResponse, error) {
+	if !r.isTableVersionCurrent(ctx) {
+		r.invalidateCachedServices()
+		if err := r.fetchServices(ctx); err != nil {
+			return nil, fmt.Errorf("error loading nexus incoming services frontend cache: %w", err)
+		}
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	service, ok := r.servicesByName[request.Name]
+	if !ok {
+		return nil, serviceerror.NewNotFound(fmt.Sprintf("frontend could not find cached service with name %v", request.Name))
+	}
+
+	return &operatorservice.GetNexusIncomingServiceResponse{Service: service}, nil
+}
+
+func (r *registry) ListNexusIncomingServices(
+	ctx context.Context,
+	request *operatorservice.ListNexusIncomingServicesRequest,
+) (*operatorservice.ListNexusIncomingServicesResponse, error) {
+	if !r.isTableVersionCurrent(ctx) {
+		r.invalidateCachedServices()
+		if err := r.fetchServices(ctx); err != nil {
+			return nil, fmt.Errorf("error loading nexus incoming services frontend cache: %w", err)
+		}
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	startIdx := 0
+	if request.NextPageToken != nil {
+		nextServiceID := string(request.NextPageToken)
+
+		startFound := false
+		startIdx, startFound = slices.BinarySearch(r.serviceIDsSorted, nextServiceID)
+
+		if !startFound {
+			return nil, serviceerror.NewFailedPrecondition("could not find service indicated by nexus list incoming services next page token")
+		}
+	}
+
+	endIdx := startIdx + int(request.PageSize)
+	if endIdx > len(r.serviceIDsSorted) {
+		endIdx = len(r.serviceIDsSorted)
+	}
+
+	var nextPageToken []byte
+	if endIdx < len(r.services) {
+		nextPageToken = []byte(r.serviceIDsSorted[endIdx])
+	}
+
+	return &operatorservice.ListNexusIncomingServicesResponse{
+		NextPageToken: nextPageToken,
+		Services:      r.services[startIdx:endIdx],
+	}, nil
+}
+
+func (r *registry) isTableVersionCurrent(ctx context.Context) bool {
+	actualTableVersion, err := r.persistence.GetNexusIncomingServicesTableVersion(ctx)
+	if err != nil {
+		return false
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+	return r.tableVersion == actualTableVersion
+}
+
+func (r *registry) fetchServices(ctx context.Context) error {
```

Comment:
I think we want a separate function for the refresh loop and the initial fetch, I'm struggling following the logic here, it's too complex, especially considering that this is called from the handlers and loading from persistence doesn't apply to the `wait` path.
You also don't want concurrent runs of `fetchServices`.